### DISCUSSION TITLE
aggregate function

### DISCUSSION THREAD

**Discussion by dongri727** [serverpod/serverpod#2011] [Q&A] [answered]
Can I use aggregate functions in serverpod?
If so, how should I write it?
I tried writing the following in the endpoint, but it does not generate properly.

class AggregatedDetails {
  List<Organisations> organisations;
  List<People> people;
  List<Categories> categories;
  // 他のデータセットもここに追加できます。

  AggregatedDetails({
    required this.organisations,
    required this.people,
    required this.categories,
    // 初期化リストもここに追加。
  });
}

// 複数のデータセットを一度のリクエストで取得する集約関数
Future<AggregatedDetails> getAggregatedDetailsByPrincipalId(Session session, {int? principalId}) async {
  if (principalId == null) {
    return AggregatedDetails(
      organisations: [],
      people: [],
      categories: [],
      // 空のリストで初期化。
    );
  }

  // 各種データ取得のための非同期関数を並行実行。
  var organisationsFuture = getOrgsByPrincipalId(session, principalId: principalId);
  var peopleFuture = getPeopleByPrincipalId(session, principalId: principalId); // この関数は仮定で設定
  var categoriesFuture = getCategoriesByPrincipalId(session, principalId: principalId); // この関数も仮定で設定

  // Futureを待つ
  var organisations = await organisationsFuture;
  var people = await peopleFuture;
  var categories = await categoriesFuture;

  // 取得したデータをAggregatedDetailsにまとめて返す
  return AggregatedDetails(
    organisations: organisations,
    people: people,
    categories: categories,
    // 他のデータも同様に。
  );
}

**Comment by SandPod** [comment]
I'm not sure I understand your question.

But assuming that you want to count the number of rows in a table that match a certain filter you can use the `db.count` method.

To count organizations that match a certain filter you should be able to write something like `Organization.db.count(session, where: (t) ....);`.

If what you actually want is serialization for the `AggregatedDetails` class you could define it as a model without a database connection and populate it in the way you are doing.

**Reply by dongri727** [reply]
Thanks for your reply.
I didn't explain the situation well enough, so I will add it.

Objective:
I'm looking to optimize how search results from multiple tables are requested from the frontend. Instead of making individual requests for each table, I want to bundle these requests together, sending one comprehensive request and receiving a consolidated return from the backend.

Current Situation:
Each table's get function is already set up with its corresponding endpoint according to the respective protocol. Currently, making individual requests from the frontend does not cause any noticeable delay. However, I'm concerned about the potential costs associated with an increase in the number of connections in the future.

Request:
I have existing code on the frontend that makes these individual requests. I'd like to adapt this process so that a single request can be made to the backend, which would then handle these queries in a grouped manner and return the aggregated results. Here's the current frontend code making individual calls:

```dart
import 'dart:core';
import 'package:acorn_client/acorn_client.dart';
import 'package:flutter/foundation.dart';
import 'package:serverpod_flutter/serverpod_flutter.dart';

import '../serverpod_client.dart';

class DetailModel extends ChangeNotifier {
  List<String> detailCatt = [];
  List<String> detailPatt =[];

  List<String> detailCountriesInv = [];
  List<String> detailPlacesInv = [];
  List<String> detailCattsInv = [];
  List<String> detailPattsInv = [];
  List<String> detailStars = [];

  List<String> detailOrgs = [];
  List<String> detailPeople = [];
  List<String> detailCategories = [];
  List<String> detailTerms = [];

  Future<void> fetchDetail(int principalId) async {
    try {
      var catt = await client.countryatts.getCattsByPrincipalId(principalId: principalId);
      detailCatt = catt.map((countryatts) => countryatts.countryatt).toList();
      var patt = await client.placeatts.getPattsByPrincipalId(principalId: principalId);
      detailPatt = patt.map((placeatts) => placeatts.placeatt).toList();

      var cinv = await client.pays.getPaysByPrincipalId(principalId: principalId);
      detailCountriesInv = cinv.map((pays) => pays.pays).toList();
      var pinv = await client.places.getPlacesByPrincipalId(principalId: principalId);
      detailPlacesInv = pinv.map((places) => places.place).toList();
      var cattsInv = await client.countryatts.getCattsInvByPrincipalId(principalId: principalId);
      detailCattsInv = cattsInv.map((countryatts) => countryatts.countryatt).toList();
      var pattsInv = await client.placeatts.getPattsInvByPrincipalId(principalId: principalId);
      detailPattsInv = pattsInv.map((placeatts) => placeatts.placeatt).toList();
      var stars = await client.stars.getStarsByPrincipalId(principalId: principalId);
      detailStars = stars.map((stars) => stars.star).toList();

      var organisations = await client.organisations.getOrgsByPrincipalId(principalId: principalId);
      detailOrgs = organisations.map((organisations) => organisations.organisation).toList();
      var people = await client.people.getPeopleByPrincipalId(principalId: principalId);
      detailPeople = people.map((people) => people.person).toList();
      var categories = await client.categories.getCategoriesByPrincipalId(principalId: principalId);
      detailCategories = categories.map((categories) => categories.category).toList();
      var terms = await client.terms.getTermsByPrincipalId(principalId: principalId);
      detailTerms = terms.map((terms) => terms.term).toList();
      notifyListeners();
    } catch (e) {
      debugPrint('Error fetching detail: $e');
    }
  }
}
```

Can you provide guidance or examples on how to consolidate these frontend requests into a single backend call that performs all the necessary queries and returns a combined result set?

**Comment by SandPod** [comment, marked as answer]
OK, there are probably multiple ways to solve this task, especially when considering how much data there will be in each table.

First, an important note: Try running any request or db call asynchronously, if possible. Regardless if you move this to the backend or not, you probably don't want to `await` the individual calls one at a time. As the code is right now, before you have fetched `terms` you have awaited all the previous requests one at a time (and right now including their network transportation time).

I'll make two suggestions here, but note that I don't know the details about your specific problem.

### DTO route with the current structure
A simple way to move this to the backend and only utilize a single endpoint is to define the `DetailModel` as a Serverpod model and then populate it in the backend.
By defining it as a Serverpod model, we will generate the class serialization and deserialization.

That means you would move the code above to an enpoint that takes the `principalId` and then change the individual endpoint calls, such as `var catt = await client.countryatts.getCattsByPrincipalId(principalId: principalId);` into the corresponding DB call that you are already doing in the endpoint.

Then, once all asynchronous calls have returned, you populate and return the `DetailModel`.

### Relational queries.
All these data types seem connected through a single `principalId`.

This opens up the possibility of doing a relational query from the `Principal` class. I suspect it would be possible to add an object relation to the principal class and then simply use includes for the different connected data fields you would like to find.

You can read more about it here:
- Defining bidirectional one-to-many relations: https://docs.serverpod.dev/concepts/database/relations/one-to-many#bidirectional-relations
- Include data from list relation: https://docs.serverpod.dev/concepts/database/relation-queries#include-relational-lists

And if you want some examples here are tests for including list relations: https://github.com/serverpod/serverpod/blob/main/tests/serverpod_test_server/test_integration/database_operations/entity_relations/entities_with_list_relations/find/list_include_test.dart

This would simplify the query to something like this:
```dart
Principal.db.findById(
  session,
  principalId,
  include: Principal.include(
    cats: Cats.includeList(),
    patts: Patts.includeList(),
    // Add the rest....
  ),
);
```

Note that this will include all the data in the `Principal` class along with all data in the `Cats` class. So, returning the retrieved object directly means that that data will be transferred to the client as well (serialization, deserialization and network transport cost).

**Reply by dongri727** [reply]
Thanks for the advice.
I will reconsider a bit more fundamentally, starting with the TABLE structure.
I created tables one after another on the spur of the moment.
